File: crunchy-core/src/main/java/io/crunchy/core/stage/NoiseFilterStage.java
package io.crunchy.core.stage;

import io.crunchy.core.pipeline.PipelineStage;
import io.crunchy.core.pipeline.ProcessingContext;
import io.crunchy.core.pipeline.SourceText;
import io.crunchy.core.text.TextUtil;

import java.util.regex.Pattern;

/**
 * Removes whole source texts that carry no engineering signal: bot chatter,
 * greetings/acknowledgements, emoji spam and raw log dumps.
 */
public final class NoiseFilterStage implements PipelineStage {

    private static final Pattern BOT_AUTHOR = Pattern.compile(
            "(?i)\\b(bot|jenkins|automation|pipeline|github-actions|dependabot|sonar)\\b");
    private static final Pattern BOT_BODY = Pattern.compile(
            "(?i)^(build\\s+#?\\d+\\s+(started|succeeded|failed|passed)|deployment\\s+(started|completed|triggered)|branch\\s+\\S+\\s+was\\s+(created|deleted)).*");
    private static final Pattern GREETING_ONLY = Pattern.compile(
            "(?i)^(?:(?:hi|hello|hey)(?: (?:team|all|everyone|folks|guys))?|thanks|thank you|thx|ty|\\+1|lgtm|ok|okay|sounds good|got it|noted|any update\\??|ping|bump)[\\s!.,]*$");
    private static final Pattern EMOJI_OR_SYMBOL = Pattern.compile(
            "[\\p{So}\\p{Sk}\\p{Cs}:;()\\[\\]{}\\-_*+~^\\s]+");

    @Override
    public String name() {
        return "noise-filter";
    }

    @Override
    public void process(ProcessingContext context) {
        int bots = 0;
        int greetings = 0;
        int logDumps = 0;
        int emoji = 0;
        for (SourceText source : context.getSources()) {
            if (source.isRemoved() || source.getOrigin().equals("description")) {
                continue;
            }
            String text = source.getText();
            if (text.isBlank()) {
                source.remove("empty");
                continue;
            }
            if (source.isBot() || BOT_AUTHOR.matcher(source.getAuthor()).find() || BOT_BODY.matcher(text.strip()).matches()) {
                source.remove("bot message");
                bots++;
            } else if (GREETING_ONLY.matcher(text.strip()).matches()) {
                source.remove("greeting/acknowledgement");
                greetings++;
            } else if (EMOJI_OR_SYMBOL.matcher(text).matches()) {
                source.remove("emoji/symbol spam");
                emoji++;
            } else if (TextUtil.looksLikeLogDump(text)) {
                source.remove("log dump");
                logDumps++;
                context.getIgnoredContent().add(
                        "Log/stack-trace dump removed from " + source.getOrigin()
                                + " (" + text.length() + " chars)");
            }
        }
        if (bots > 0) {
            context.getIgnoredContent().add(bots + " bot message(s) removed");
        }
        if (greetings > 0) {
            context.getIgnoredContent().add(greetings + " greeting/acknowledgement comment(s) removed");
        }
        if (emoji > 0) {
            context.getIgnoredContent().add(emoji + " emoji-only comment(s) removed");
        }
    }
}

File: web/src/core/stage/NoiseFilterStage.ts
import { PipelineStage } from '../pipeline/PipelineStage.js';
import { ProcessingContext } from '../pipeline/ProcessingContext.js';
import { looksLikeLogDump } from '../text/TextUtil.js';

/**
 * Removes whole source texts that carry no engineering signal: bot chatter,
 * greetings/acknowledgements, emoji spam and raw log dumps.
 */
const BOT_AUTHOR = /\b(bot|jenkins|automation|pipeline|github-actions|dependabot|sonar)\b/i;
const BOT_BODY =
  /^(build\s+#?\d+\s+(started|succeeded|failed|passed)|deployment\s+(started|completed|triggered)|branch\s+\S+\s+was\s+(created|deleted)).*$/i;
const GREETING_ONLY =
  /^(?:(?:hi|hello|hey)(?: (?:team|all|everyone|folks|guys))?|thanks|thank you|thx|ty|\+1|lgtm|ok|okay|sounds good|got it|noted|any update\??|ping|bump)[\s!.,]*$/i;
const EMOJI_OR_SYMBOL = /^[\p{So}\p{Sk}\p{Cs}:;()[\]{}\-_*+~^\s]+$/u;

export class NoiseFilterStage implements PipelineStage {
  readonly name = 'noise-filter';

  process(context: ProcessingContext): void {
    let bots = 0;
    let greetings = 0;
    let logDumps = 0;
    let emoji = 0;
    for (const source of context.sources) {
      if (source.removed || source.origin === 'description') {
        continue;
      }
      const text = source.text;
      if (text.trim().length === 0) {
        source.remove('empty');
        continue;
      }
      const trimmed = text.trim();
      if (source.bot || BOT_AUTHOR.test(source.author) || BOT_BODY.test(trimmed)) {
        source.remove('bot message');
        bots++;
      } else if (GREETING_ONLY.test(trimmed)) {
        source.remove('greeting/acknowledgement');
        greetings++;
      } else if (EMOJI_OR_SYMBOL.test(text)) {
        source.remove('emoji/symbol spam');
        emoji++;
      } else if (looksLikeLogDump(text)) {
        source.remove('log dump');
        logDumps++;
        context.ignoredContent.push(
          `Log/stack-trace dump removed from ${source.origin} (${text.length} chars)`,
        );
      }
    }
    if (bots > 0) {
      context.ignoredContent.push(`${bots} bot message(s) removed`);
    }
    if (greetings > 0) {
      context.ignoredContent.push(`${greetings} greeting/acknowledgement comment(s) removed`);
    }
    if (emoji > 0) {
      context.ignoredContent.push(`${emoji} emoji-only comment(s) removed`);
    }
    void logDumps;
  }
}
